src/sfizz/Voice.cpp

```diff
@@ -918,63 +956,85 @@ void Voice::Impl::fillWithData(AudioSpan<float> buffer) noexcept
+
+    const auto addPartitionIfNecessary = [&] (unsigned blockIndex, int wrappedIndex, bool wrapped) {
+        const bool xfading = wrappedIndex >= loop.start && wrappedIndex >= loop.xfOutStart;
+        const int partitionType = xfading ? kPartitionLoopXfade : kPartitionNormal;
+
+        // if looping or entering a different type, start a new partition
+        bool start = blockIndex == 0 || wrapped || partitionType != oldPartitionType;
+        if (start) {
+            partitionStarts[numPartitions] = blockIndex;
+            partitionTypes[numPartitions] = partitionType;
+            ++numPartitions;
+        }
+
+        oldIndex = wrappedIndex;
+        oldPartitionType = partitionType;
+    };
 
-            // wrap indices post loop-entry around the loop segment
-            int wrappedIndex = (index <= loop.end) ? index :
-                (loop.start + (index - loop.start) % loop.size);
+    if (shouldLoop) {
+        unsigned i = 0;
+        while (i < numSamples) {
+            int wrappedIndex = (*indices)[i] - loop.size * blockRestarts;
+            if (wrappedIndex > loop.end) {
+                wrappedIndex -= loop.size;
+                blockRestarts += 1;
+                loop_.restarts += 1;
+            }
             (*indices)[i] = wrappedIndex;
+            const bool wrapped = wrappedIndex < oldIndex;
 
             // identify the partition this index is in
-            bool xfading = wrappedIndex >= loop.start && wrappedIndex >= loop.xfOutStart;
-            int partitionType = xfading ? kPartitionLoopXfade : kPartitionNormal;
-            // if looping or entering a different type, start a new partition
-            bool start = i == 0 || wrappedIndex < oldIndex || partitionType != oldPartitionType;
-            if (start) {
-                partitionStarts[numPartitions] = i;
-                partitionTypes[numPartitions] = partitionType;
-                ++numPartitions;
-            }
+            addPartitionIfNecessary(i, wrappedIndex, wrapped);
+            i++;
 
-            oldIndex = wrappedIndex;
-            oldPartitionType = partitionType;
+            // Break if we reached the loop count
+            if (wrapped && region_->loopCount && loop_.restarts >= *region_->loopCount)
+                break;
         }
-    }
-    // index preprocessing for one-shots
-    else {
-        // cut short the voice at the instant of reaching end of sample
-        const auto sampleEnd = min(
-            int(region_->trueSampleEnd(resources_.filePool.getOversamplingFactor())),
-            int(currentPromise_->information.end),
-            int(source.getNumFrames()))
-            - 1;
 
+        while (i < numSamples) { // In case we released within the block, continue as if it were a one-shot
+            (*indices)[i] -= loop.size * blockRestarts;
+            if ((*indices)[i] >= sampleEnd) {
+                fill<int>(indices->subspan(i), sampleEnd);
+                fill<float>(coeffs->subspan(i), 1.0f);
```

Comment:
Reduce the value of coeff into domain [0;1[
```suggestion
                fill<float>(coeffs->subspan(i), std::nextafter(1.0f, 0.0f));
```